demes/demes.py

```diff
@@ -1937,3 +2036,196 @@ def collapse_demes(pairs):
         simplify_epochs(data)
 
         return data
+
+
+class Builder:
+    """
+    The Builder provides a set of convenient methods for incrementally
+    constructing a deme graph. The state of the graph is stored internally as a
+    dictionary of objects following Demes' data model, and may be converted
+    into a fully-resolved :class:`Graph` object using the :meth:`.resolve()` method.
+
+    TODO: add some example code.
+
+    :ivar dict data: The data dictionary of the graph's current state.
+        The objects nested within this dictionary follow Demes' data model,
+        as described in the :ref:`schema <sec_schema>`.
+
+        .. note::
+            Users may freely modify the data dictionary, as long as the data
+            model is not violated.
+    """
+
+    def __init__(
+        self,
+        *,
+        description: str = None,
+        time_units: str = "generations",
+        generation_time: float = None,
+        doi: list = None,
+        defaults: dict = None,
+    ):
+        """
+        :param str description: A human readable description of the demography.
+            May be ``None``.
+        :param str time_units: The units of time used for the demography. This is
+            commonly ``years`` or ``generations``, but can be any string.
+        :param float generation_time: The generation time of demes, in units given
+            by the ``time_units`` parameter. Concretely, dividing all times
+            by ``generation_time`` will convert the graph to have time
+            units in generations.  If ``generation_time`` is ``None``, the units
+            are assumed to be in generations already.
+        :param doi: If the graph describes a published demography, the DOI(s)
+            should be be given here as a list.
+        :vartype doi: list of str
+        """
+        self.data: MutableMapping[str, Any] = dict(time_units=time_units)
```

Comment:
Maybe add the lists here, to simplify things a bit later? E.g, ``dict(time_units=time_units, demes=[], migrations=[]...)``

Reply:
I had that at some point, but figured that having the explicit check and list construction inside `add_deme()` provides slightly greater robustness/flexibility in the case when someone uses `Builder.fromdict()`.